src/dbnode/storage/index.go

```diff
@@ -1429,16 +1429,15 @@ func (i *nsIndex) query(
 			i.metrics.queryNonExhaustiveLimitError.Inc(1)
 		}
 
-		err := fmt.Errorf(
+		// NB(r): Make sure error is not retried and returns as bad request.
+		return exhaustive, xerrors.NewInvalidParamsError(fmt.Errorf(
```

Comment:
should we introduce a new error type so we can return a more relevant error code, like 429 (Too Many Requests). That way a client can know they should backoff, instead of thinking the request was invalid.